language-plutus-core/src/Language/PlutusCore/Error.hs

```diff
@@ -43,10 +43,17 @@ data RenameError a
     deriving (Show, Eq, Generic, NFData)
 
 data TypeError a
-    = KindMismatch a (Type TyNameWithKind ()) (Kind ()) (Kind ())
-    | TypeMismatch a (Term TyNameWithKind NameWithType ())
-                     (Type TyNameWithKind ())
-                     (NormalizedType TyNameWithKind ())
+    = NotImplemented
+    | KindMismatch { _loc          :: a
+                   , _inType       :: Type TyNameWithKind ()
+                   , _expectedKind :: Kind ()
+                   , _foundKind    :: Kind ()
+                   }
+    | TypeMismatch { _loc          :: a
+                   , _inTerm       :: Term TyNameWithKind NameWithType ()
+                   , _expectedType :: Type TyNameWithKind ()
+                   , _foundType    :: NormalizedType TyNameWithKind ()
+                   }
```

Comment:
Is it acceptable to not export the fields and still allow them for internal documentation? That was my main purpose (since Haddock will generate docs for such things).

Reply:
Haddock will generate docs for things like

```haskell
KindMismatch
    a  -- ^ Annotation.
    Type TyNameWithKind () -- ^ In type.
    (Kind ()) -- ^ Expected kind.
    (Kind ()) -- ^ Found kind.
```

no?